pontoon/projects/views.py

```diff
@@ -62,6 +71,90 @@ def ajax_info(request, slug):
     })
 
 
+@permission_required_or_403('base.can_manage_project')
+@transaction.atomic
+@require_AJAX
+def ajax_notifications(request, slug):
+    """Notifications tab."""
+    project = get_object_or_404(Project.objects.available(), slug=slug)
+    available_locales = project.locales.order_by('name')
+
+    # Send notifications
+    if request.method == 'POST':
+        message = request.POST.get('message', None)
+        selected_ids = request.POST.get('selected_locales')
+
+        # TODO: implement validation on client side
+        if not selected_ids:
+            return HttpResponseBadRequest(
+                'no_locales'
+            )
+
+        if not message:
+            return HttpResponseBadRequest(
+                'no_message'
+            )
+
+        contributors = User.objects.filter(
+            translation__entity__resource__project=project,
+        )
+
+        # For performance reasons, only filter contributors for selected
+        # locales if different from all project locales
+        available_ids = sorted(list(available_locales.values_list('id', flat=True)))
+        selected_ids = sorted(map(int, selected_ids.split(',')))
+
+        if available_ids != selected_ids:
+            contributors = User.objects.filter(
+                translation__entity__resource__project=project,
+                translation__locale__in=available_locales.filter(id__in=selected_ids)
+            )
+
+        identifier = uuid.uuid4().hex
+        for contributor in contributors.distinct():
+            notify.send(
+                request.user,
```

Comment:
I have benchmarks. :) We've been using the same code for automated notifications and it performs good enough to be used via `XMLHttpRequest`, too.

The slow part is actually retrieving the list of contributors, which is also hitting us on /contributors pages. We should figure out the way to optimize it.

But in general I agree - we should defer this to celery if the requests start timing out.